### PR TITLE
nixos/mailpit: init

### PR DESCRIPTION
**Pull request by Ma27**
## Description of changes
Simple module to configure a systemd service for mailpit.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/mail/mailpit.nix
Comment (leona-ya) [inline nixos/modules/services/mail/mailpit.nix:42]: This seems to be outdated, but as nixpkgs has a quite old version of mailpit it seems correct for now..

**Review by leona-ya — APPROVED** [review]

File: nixos/modules/services/mail/mailpit.nix
Comment (Ma27) [reply nixos/modules/services/mail/mailpit.nix:42]: Oof I already wondered why how I came up with `database` while hacking on that module 🤦 
I guess I'll do an update of the software then, thanks for catching this.

Comment (RafaelKr) [reply nixos/modules/services/mail/mailpit.nix:42]: FYI I already opened https://github.com/NixOS/nixpkgs/pull/324302 for automatic update PRs via @r-ryantm 
Still waiting to get it merged.

Btw I also initiated that the mailpit documentation shows the version in which specific options were added. See https://github.com/axllent/mailpit/issues/328 for more info.

**Comment by RafaelKr** [comment]
May I suggest to make it possible to configure multiple servers? My use case is to have multiple environments (dev, staging, production) on one host and I like to have an instance for each.

I already have built something for a local project but didn't have time to work it out with all options and create a PR for it.

But here's my implementation: https://gist.github.com/RafaelKr/e9ac5850da8b674a92f0bf4bb5bb9b7f

**Comment by Ma27** [comment]
ping @leona-ya @RafaelKr 
Just rebased against latest master.
PR should be ready to merge I guess? :)

**Comment by RafaelKr** [comment]
I think mailpit listens on `0.0.0.0` by default to make it easily usable inside docker. For usage with nix I would suggest switching to the more secure `127.0.0.1` by default. What do you think?
Besides from that it LGTM! :)

Maybe we want to wait for the merge of #334352 and rebase again before merging this PR.

File: nixos/modules/services/mail/mailpit.nix
Comment (RafaelKr) [inline nixos/modules/services/mail/mailpit.nix:62]: Maybe default to `127.0.0.1:8025`. See https://github.com/NixOS/nixpkgs/pull/329657#issuecomment-2288699086

File: nixos/modules/services/mail/mailpit.nix
Comment (RafaelKr) [inline nixos/modules/services/mail/mailpit.nix:69]: Maybe default to `127.0.0.1:1025`. See https://github.com/NixOS/nixpkgs/pull/329657#issuecomment-2288699086

**Review by RafaelKr — CHANGES_REQUESTED** [review]

**Comment by Ma27** [comment]
> Maybe we want to wait for the merge of #334352 and rebase again before merging this PR.

This shouldn't give a conflict: `source.nix` isn't touched anymore in this PR.

**Comment by RafaelKr** [comment]
Let's go! 🚀 :)

**Comment by RafaelKr** [comment]
This is from GSOC 2026 ideas: https://gsoc-explorer.vercel.app/ideas/nixos-foundation/#migrate-some-nixos-modules-to-modular-services

> [Modular services](https://nixos.org/manual/nixos/stable/#modular-services)
are a new NixOS feature to reduce evaluation time as more service modules are written, contrasting the historic evaluation time increases as the [central module list](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/module-list.nix)
increased, while also supporting more flexible usage patterns like multiple instances of a service type on the same host.
> The workflow:
>
> - Pick a NixOS service whose maintainers agree to the pilot project
> - Factor the core of the service into a modular services module, keeping the user-facing options of the NixOS module the same
> - Implement a very basic test for the modular service itself that only tests the functionality of the modular service (typically config file and launch), leaving proper integration tests to the NixOS test

@Ma27 with the instances option this would now be a perfect candidate for a 1:1 port as a modular service, WDYT?

**Pull request merged**
